packages/fdc3-get-agent/src/strategies/PostMessageLoader.ts

```diff
@@ -0,0 +1,160 @@
+import { DesktopAgent, GetAgentParams } from '@kite9/fdc3-standard'
+import { createDesktopAgentAPI } from '../messaging/message-port';
+import { v4 as uuidv4 } from "uuid"
+import { ConnectionDetails } from '../messaging/MessagePortMessaging';
+import { Loader } from './Loader';
+import { BrowserTypes } from "@kite9/fdc3-schema";
+import { FDC3_VERSION } from './getAgent';
+
+type WebConnectionProtocol1Hello = BrowserTypes.WebConnectionProtocol1Hello
+type WebConnectionProtocol2LoadURL = BrowserTypes.WebConnectionProtocol2LoadURL
+type WebConnectionProtocol3Handshake = BrowserTypes.WebConnectionProtocol3Handshake
+
+function collectPossibleTargets(w: Window, found: Window[]) {
+    if (w) {
+        if (found.indexOf(w) == -1) {
+            found.push(w);
+        }
+
+        if (found.indexOf(w.opener) == -1) {
+            collectPossibleTargets(w.opener, found);
+        }
+
+        if (found.indexOf(w.parent) == -1) {
+            collectPossibleTargets(w.parent, found);
+        }
+    }
+}
+
+
+/**
+ * Starts the connection process off by sending a hello message
+ */
+function sendWCP1Hello(w: MessageEventSource, options: GetAgentParams, connectionAttemptUuid: string, origin: string) {
+    const requestMessage: WebConnectionProtocol1Hello = {
+        type: 'WCP1Hello',
+        meta: {
+            connectionAttemptUuid: connectionAttemptUuid,
+            timestamp: new Date()
+        },
+        payload: {
+            channelSelector: options.channelSelector,
+            fdc3Version: FDC3_VERSION,
+            resolver: options.intentResolver,
+            identityUrl: options.identityUrl!!,
+            actualUrl: globalThis.window.location.href
+        }
+    }
+
+    w.postMessage(requestMessage, { targetOrigin: origin });
+}
+
+/**
+ * The desktop agent requests that the client opens a URL in order to provide a message port.
+ */
+function openFrame(url: string): Window {
+    const IFRAME_ID = "fdc3-communications-embedded-iframe"
+
+    // remove an old one if it's there
+    const existing = document.getElementById(IFRAME_ID)
+    if (existing) {
+        existing.remove()
+    }
+
+    // create a new one
+    var ifrm = document.createElement("iframe")
+    ifrm.setAttribute("src", url)
+    ifrm.setAttribute("id", IFRAME_ID)
+    ifrm.setAttribute("name", "FDC3 Communications")
+    ifrm.style.width = "0px"
+    ifrm.style.height = "0px"
+    ifrm.style.border = "0"
+    ifrm.style.position = "fixed"
+    document.body.appendChild(ifrm)
+    return ifrm.contentWindow!!
+}
+
+export function helloExchange(options: GetAgentParams, connectionAttemptUuid: string): Promise<ConnectionDetails> {
```

Comment:
What happens to this method if data.type doesn't equal `WCP3Handshake`? It looks like it won't resolve otherwise... do you just let it hang and timeout? What's the expected interaction?